File: Templates/floyds_cycle_finding.cpp
#include <cstddef>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

bool hasCycle(ListNode* head) {
    if (head == nullptr) return false;
    ListNode* slow = head;
    ListNode* fast = head;
    while (true) {
        if (fast == nullptr || fast->next == nullptr) {
            return false;
        }
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) {
            break;
        }
    }
    return true;
}

File: Templates/floyds_cycle_finding.py
def hasCycle(head: Optional[ListNode]) -> bool:
    if head is None:
        return False
    
    slow = fast = head

    while True :
        if fast is None or fast.next is None:
            return False

        slow = slow.next
        fast = fast.next.next

        if slow == fast:
            break

    return True